crates/core/component/auction/src/component/dutch_auction.rs

```diff
@@ -51,11 +63,147 @@ pub(crate) trait DutchAuctionManager: StateWrite {
+            .expect("the auction exists");
+
+        let DutchAuctionDescription {
+            input,
+            output_id,
+            max_output: _,
+            min_output: _,
+            start_height,
+            end_height,
+            step_count,
+            nonce: _,
+        } = old_dutch_auction.description;
+
+        let DutchAuctionState {
+            sequence: _,
+            current_position,
+            next_trigger: _,
+            input_reserves: _,
+            output_reserves: _,
+        } = old_dutch_auction.state;
+
+        let auction_input_id = input.asset_id;
+        let auction_output_id = output_id;
+
+        let auction_trigger = TriggerData {
+            start_height,
+            end_height,
+            step_count,
+        };
+
+        // Recover the LP's balances, if it exists.
+        let lp_reserves = if let Some(auction_lp_id) = current_position {
+            self.close_position_by_id(&auction_lp_id)
+                .await
+                .map_err(|e| {
+                    tracing::error!(
+                        ?e,
+                        ?auction_lp_id,
+                        ?auction_id,
+                        "failed to close dutch auction LP"
+                    )
+                })
+                .expect("position should exist and be opened or closed");
+            self.withdraw_position(auction_lp_id, 0u64)
+                .await
+                .map_err(|e| {
+                    tracing::error!(
+                        ?e,
+                        ?auction_lp_id,
+                        ?auction_id,
+                        "failed to close dutch auction LP"
+                    )
+                })
+                .expect("no state incoherence")
+        } else {
+            Balance::zero()
+        };
+
+        // We remove the execution trigger that we are currently processing:
+        self.unset_trigger_for_dutch_id(auction_id, trigger_height);
+
+        // Prepare a new auction, based on the previous one.
+        let mut new_dutch_auction = DutchAuction {
+            description: old_dutch_auction.description,
+            state: old_dutch_auction.state,
+        };
+
+        let balance_input_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == input.asset_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
+        let balance_output_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == output_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
+
+        // After consuming the LP, we reset the state, getting ready to either
+        // execute another session, or retire the auction.
+        new_dutch_auction.state.current_position = None;
+        new_dutch_auction.state.input_reserves = balance_input_asset;
+        new_dutch_auction.state.output_reserves += balance_output_asset;
+        new_dutch_auction.state.next_trigger = 0;
```

Comment:
This is a little confusing, I think it would be cleaner to model the data as `next_trigger: Option<NonZeroU64>` so that we can have `Some(height)` or `None` in our code and we don't have to think about the 0 special case (even though `Option<NonZeroU64>` will specialize to exactly the same representation).